File: Competitive Programming 3/Section 1.4/Anagram/642 - Word Amalgamation/Java/Main.java
import java.util.Scanner;
import java.util.HashMap;
import java.util.Vector;
import java.util.Arrays;
import java.util.Map;
import java.util.Collections;

public class Main{
    static Scanner sc = new Scanner(System.in);
    static HashMap<String,Vector<String>> anagrams = new HashMap<String,Vector<String>>();
    public static void main(String[]args){
        String line;
        while((line=sc.next()).charAt(0)!='X'){
            char[] lineC = line.toCharArray();
            Arrays.sort(lineC);
            String sortedLine = new String(lineC);
            if(anagrams.containsKey(sortedLine))
                anagrams.get(sortedLine).add(line);
            else{
                final String lineCp = line;
                anagrams.put(sortedLine,new Vector<String>(){
                    {
                        add(lineCp);
                    }
                });
            }
        }
        //sort Vectors
        anagrams.forEach((k,v) -> Collections.sort(v));
        //answer anagram queries
        while((line=sc.next()).charAt(0)!='X'){
            char[] lineC = line.toCharArray();
            Arrays.sort(lineC);
            line = new String(lineC);
            if(anagrams.containsKey(line))
                for(String element: anagrams.get(line))
                    System.out.println(element);
            else
                System.out.println("NOT A VALID WORD");
            System.out.println("******");
        }
    }
}

File: Competitive Programming 3/Section 1.4/Anagram/642 - Word Amalgamation/Python/main.py
anagrams = {}
def main():
    while True:
        line = input()
        if line[0] == 'X':
            break
        sortedLine = ''.join(sorted(line))
        if sortedLine not in anagrams:
            anagrams[sortedLine] = []
        anagrams[sortedLine].append(line)
    #sort vectors
    for an in anagrams.values():
        an.sort()
    #answer anagram queries
    while True:
        line = input()
        if line[0] == 'X':
            break
        line = ''.join(sorted(line))
        if line in anagrams:
            for element in anagrams[line]:
                print(element)
        else:
            print('NOT A VALID WORD')
        print('******')
main()